From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Anoob Soman <anoob.soman@citrix.com>
Date: Tue, 9 May 2017 17:13:28 +0100
Subject: [PATCH] CA-247413: Make pxebs accept broadcast DHCP packets

In start_pxebs() (net/udp/dhcp.c), before doing xfer_open_socket(),
fetch_ipv4_setting() makes sure that IP address (assigned during DHCP
process) is set to dchp->local.sin_addr. This means, after DGRAM socket
is opened, dhcp_pxebs_rx() cannot receive any broadcasted DHCP ACK
packets. udp_demux will allow only packets unicasted to interface's IP
address, to this open UDP connection. This breaks pxebs.

Move fetch_ipv4_setting() after xfer_open_socket(), such that DGRAM
socket is opened with dchp->local.sin_addr set to zero, (just like
start_dhcp()) and udp_demux will let broacast DHCP packets during pxebs.

Ported to v2.0.0 from XS9 ipxe 20121005-1.0.8.

Signed-off-by: Tu Dinh <ngoc-tu.dinh@vates.tech>

diff --git a/src/net/udp/dhcp.c b/src/net/udp/dhcp.c
index 59b7c663d24a..d608cf2af53b 100644
--- a/src/net/udp/dhcp.c
+++ b/src/net/udp/dhcp.c
@@ -1477,8 +1477,6 @@ int start_pxebs ( struct interface *job, struct net_device *netdev,
 	timer_init ( &dhcp->timer, dhcp_timer_expired, &dhcp->refcnt );
 	dhcp->netdev = netdev_get ( netdev );
 	dhcp->local.sin_family = AF_INET;
-	fetch_ipv4_setting ( netdev_settings ( netdev ), &ip_setting,
-			     &dhcp->local.sin_addr );
 	dhcp->local.sin_port = htons ( BOOTPC_PORT );
 	dhcp->pxe_type = cpu_to_le16 ( pxe_type );
 
@@ -1527,6 +1525,8 @@ int start_pxebs ( struct interface *job, struct net_device *netdev,
 				  ( struct sockaddr * ) &dhcp->local ) ) != 0 )
 		goto err;
 
+	fetch_ipv4_setting ( netdev_settings ( netdev ), &ip_setting,
+			     &dhcp->local.sin_addr );
 	/* Enter PXEBS state */
 	dhcp_set_state ( dhcp, &dhcp_state_pxebs );
 
